dev/io.openliberty.microprofile.faulttolerance.telemetry/src/io/openliberty/microprofile/faulttolerance/telemetry/metrics/integration/AbstractMetricRecorderImpl.java

```diff
@@ -0,0 +1,474 @@
+        } else {
+            ftCircuitbreakerCallsTotal = null;
+            ftCircuitbreakerStateTotal = null;
+            ftCircuitbreakerOpenedTotal = null;
+        }
+
+        if (bulkheadPolicy != null) {
+            ftBulkheadCallsTotal = meter.counterBuilder(metricPrefix + ".bulkhead.calls.total").setDescription(Tr.formatMessage(tc, "ft.bulkhead.calls.total.description")).build();
+            ftBulkheadExecutionsRunning = meter.gaugeBuilder(metricPrefix + ".bulkhead.executionsRunning").ofLongs().setDescription(Tr.formatMessage(tc, "ft.bulkhead.executionsRunning.description")).buildWithCallback(this::getConcurrentExecutions);
+            ftBulkheadRunningDuration = meter.histogramBuilder(metricPrefix + ".bulkhead.runningDuration").setDescription(Tr.formatMessage(tc, "ft.bulkhead.runningDuration.description")).ofLongs().setUnit("nanoseconds").build();
+
+            if (isAsync == AsyncType.ASYNC) {
+                ftBulkheadExecutionsWaiting = meter.gaugeBuilder(metricPrefix + ".bulkhead.executionsWaiting").ofLongs().setDescription(Tr.formatMessage(tc, "ft.bulkhead.executionsWaiting.description")).buildWithCallback(this::getQueuePopulation);
+                ftBulkheadWaitingDuration = meter.histogramBuilder(metricPrefix + ".bulkhead.waitingDuration").setDescription(Tr.formatMessage(tc, "ft.bulkhead.waitingDuration.description")).ofLongs().setUnit("nanoseconds").build();
+            } else {
+                ftBulkheadExecutionsWaiting = null;
+                ftBulkheadWaitingDuration = null;
+            }
+        } else {
+            ftBulkheadCallsTotal = null;
+            ftBulkheadExecutionsRunning = null;
+            ftBulkheadExecutionsWaiting = null;
+            ftBulkheadRunningDuration = null;
+            ftBulkheadWaitingDuration = null;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Trivial
+    @Override
+    public void incrementInvocationSuccessCount(FallbackOccurred fallbackOccurred) {
+        incrementFtTotalCount(fallbackOccurred, true);
+    }
+
+    /** {@inheritDoc} */
+    @Trivial
+    @Override
+    public void incrementInvocationFailedCount(FallbackOccurred fallbackOccurred) {
+        incrementFtTotalCount(fallbackOccurred, false);
+    }
+
+    @Trivial
+    private void incrementFtTotalCount(FallbackOccurred fallbackOccurred, boolean invocationSucceeded) {
+        AttributesBuilder ab = Attributes.builder().putAll(methodAttribute);
+
+        if (!fallbackDefined) {
+            ab.put("fallback", "notDefined");
+        } else if (fallbackOccurred == FallbackOccurred.WITH_FALLBACK) {
+            ab.put("fallback", "applied");
+        } else {
+            ab.put("fallback", "notApplied");
+        }
+
+        if (invocationSucceeded) {
+            ab.put("result", "valueReturned");
+        } else {
+            ab.put("result", "exceptionThrown");
+        }
+
+        if (ftInvocationsTotal != null) {
+            ftInvocationsTotal.add(1, ab.build());
```

Comment:
I note that [add(T, Annotations)](https://www.javadoc.io/static/io.opentelemetry/opentelemetry-api/1.34.0/io/opentelemetry/api/metrics/LongCounter.html#add(long,io.opentelemetry.api.common.Attributes)) may call `Context.current()`.

Since we will usually update several metrics at once, we might see a performance improvement if we were to call `Context.current()` once and pass it through. Not sure of the overhead of `Context.current()` but I think it uses a `ThreadLocal`.

Unfortunately, we'd need to pass that through from FT code, and provide an abstraction because MP Metrics has no concept of `Context`.

For the moment, I think we shouldn't change this unless we find it causes performance issues, but it's something to look out for in future.

Reply:
On a similar note, we might see performance issues from building the `Attributes` object anew each request, rather than using the same one for each method call. Again, **I wouldn't change this now** since it will make the code more complex, but we might have to do it if performance testing shows issues.